torch/nn/parallel/distributed.py

```diff
@@ -563,6 +563,12 @@ def forward(self, *inputs, **kwargs):
         if self.require_forward_param_sync:
             self._sync_params()
 
+        # It may allocate new buckets before deallocating old buckets
+        # inside rebuildBuckets.
+        # To save peak memory usage, call rebuildBuckets before the peak
+        # memory usage increases during computing outputs.
```

Comment:
Maybe I'm missing something here, but in `initialize_buckets` it seems like we first clear the existing buckets before allocating new ones? (https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/reducer.cpp#L624).

Reply:
right, it cleared existing buckets without this diff. With this diff, variable.grad is pointing to the bucket tensor as well, so clear buckets_ will not clear the existing buckets